test/integration/cds_integration_test.cc

```diff
@@ -0,0 +1,211 @@
+          bootstrap.mutable_dynamic_resources()->mutable_cds_config()->mutable_api_config_source();
+      bootstrap.mutable_dynamic_resources()->clear_ads_config();
+      bootstrap.mutable_dynamic_resources()->clear_lds_config();
+
+      cds_config->set_api_type(envoy::api::v2::core::ApiConfigSource::GRPC);
+      cds_config->mutable_request_timeout()->set_seconds(1);
+      auto* grpc_service = cds_config->add_grpc_services();
+      setGrpcService(*grpc_service, "my_cds_cluster", fake_upstreams_[0]->localAddress());
+      grpc_service->mutable_envoy_grpc()->set_cluster_name("my_cds_cluster");
+
+      auto* cds_cluster = bootstrap.mutable_static_resources()->add_clusters();
+      cds_cluster->set_name("my_cds_cluster");
+      cds_cluster->mutable_connect_timeout()->set_seconds(5);
+      auto* sockaddr = cds_cluster->add_hosts()->mutable_socket_address();
+      sockaddr->set_protocol(envoy::api::v2::core::SocketAddress::TCP);
+      sockaddr->set_address(Network::Test::getLoopbackAddressString(ipVersion()));
+      sockaddr->set_port_value(0);
+      cds_cluster->clear_http2_protocol_options();
+    });
+    setUpstreamProtocol(FakeHttpConnection::Type::HTTP2);
+
+    HttpIntegrationTest::initialize();
+
+    fake_upstreams_[0]->set_allow_unexpected_disconnects(false);
+    // Causes cds_connection_ to be filled with a newly constructed FakeHttpConnection.
+    AssertionResult result =
+        fake_upstreams_[0]->waitForHttpConnection(*dispatcher_, cds_connection_);
+    RELEASE_ASSERT(result, result.message());
+    result = cds_connection_->waitForNewStream(*dispatcher_, cds_stream_);
+    RELEASE_ASSERT(result, result.message());
+    cds_stream_->startGrpcStream();
+
+    EXPECT_TRUE(compareDiscoveryRequest(Config::TypeUrl::get().Cluster, "", {}));
+    sendDiscoveryResponse<envoy::api::v2::Cluster>(Config::TypeUrl::get().Cluster,
+                                                   {buildCluster("cluster_0")}, "1");
+  }
+
+  testing::NiceMock<Server::Configuration::MockTransportSocketFactoryContext> factory_context_;
+  FakeStreamPtr cds_stream_;
+  FakeHttpConnectionPtr cds_connection_;
+  FakeHttpConnectionPtr upstream_connection_;
+};
+
+// GoogleGrpc causes problems.
+INSTANTIATE_TEST_CASE_P(IpVersionsClientType, CdsIntegrationTest,
+                        testing::Combine(testing::ValuesIn(TestEnvironment::getIpVersionsForTest()),
+                                         testing::Values(Grpc::ClientType::EnvoyGrpc)));
+
+// Tests Envoy HTTP health checking a single healthy endpoint and reporting that it is
+// indeed healthy to the server.
+TEST_P(CdsIntegrationTest, RouterRequestAndResponseWithBodyNoBuffer) {
+  // Controls how many fake_upstreams_.emplace_back(new FakeUpstream) will happen in
+  // BaseIntegrationTest::createUpstreams() (which is part of initialize()).
+  setUpstreamCount(1);
+  initializeCds();
+
+  // Adapted from HttpIntegrationTest::testRouterRequestAndResponseWithBody(1024, 512, false).
```

Comment:
This is pretty low-level. You should be able to get away with just calling `HttpIntegrationTest::testRouterHeaderOnlyRequestAndResponse()` for the body (or something like that).

Reply:
Oh, nice! Yes, sticking to the headers-only test sounds good. However, it still needs a change within the test to work. Specifically, lookupPort("http") has to be replaced. I figured it wasn't too bad to dump the registerTestServerPorts() / lookupPort() mechanism, since ultimately it's entirely within the test framework, and just another way of saving+recalling a port.

I think lookupPort() not working as expected is related to your other comments about the upstreams.

Reply:
Yeah, if you can add a regular upstream that looks to the rest of the test infrastructure like the standard data plane one, this should "just work".

Reply:
Unfortunately, that's the thing I've had the most problems with. Earlier, when I tried to define a regular upstream, initialize() hung forever waiting for the listener to come up. I was only ever able to fix that by giving up on a separate regular upstream. After discovering HTTP_PROXY_CONFIG just now I tried working from that, but that brought me right back to the same problem.

I tried doing nothing but calling initialize() with the test fixture constructed with that config. initialize() was able to finish; so far so good. I tried adding back in my beforehand addition of a CDS cluster in a config modifier: that brought it back to hanging. I tried instead supplying the config in a modified version of HTTP_PROXY_CONFIG, but failed the ASSERT(port_idx == ports.size()) in finalize(). Here I am pretty stumped, because the code just before it pretty clearly advances port_idx when a cluster has a "hosts" field with a "socket_address", which both of my clusters have. Looking at the logs, my CDS cluster gets processed there, but the regular upstream does not - even though it's just the cluster_0 copied directly from HTTP_PROXY_CONFIG.

But, the really weird thing is that just the addition of the config modifier will make initialize() hang - unless I don't try creating a separate regular upstream, in which case that exact same config modifier lets the test pass.

Reply:
Yeah; if we can sort this out cleanly it would be ideal, I feel that if there is a problem here we should fix it or better document how to get multiple upstreams working, this is an important feature of the test framework. Thanks for keeping at it.

Reply:
I think I now fully understand the problem I have been having with the test framework's port map machinery. registerTestServerPorts() needs there to be listeners in order to register anything, but I think the only way I've been able to even get beyond the framework's initialization phase has been to not have listeners. Which makes me realize: when I've had the test passing, it's been by having the HTTP request to the regular upstream go directly to that upstream - not through Envoy! So, the portion of the CDS logic where Envoy actually does something with the DiscoveryResponse wouldn't be tested. Definitely has to be fixed.

Unfortunately that means I'm back to the "framework refuses to complete initialization when there is a listener" issue, which still seems completely mysterious.

Reply:
Yeah, definitely needs listeners. Also, maybe verifying with some stats that Envoy has been doing serving could an idea ;) I think we'd be open to any generalization/cleanup of the port registration and lookup that is needed. 

Also, maybe chat with @AndresGuedez about this a bit; he will need to be doing some e2e work for Scoped RDS, so having a per-xDS gRPC integration test would be handy.